include/ibamr/ComplexFluidForcing.h

```diff
@@ -0,0 +1,330 @@
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of The University of North Carolina nor the names of
+//      its contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef included_ComplexFluidForcing
+#define included_ComplexFluidForcing
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "IBAMR_config.h"
+
+#include "ibamr/AdvDiffComplexFluidConvectiveOperator.h"
+#include "ibamr/AdvDiffSemiImplicitHierarchyIntegrator.h"
+#include "ibamr/CFGiesekusRelaxation.h"
+#include "ibamr/CFOldroydBRelaxation.h"
+#include "ibamr/CFRoliePolyRelaxation.h"
+#include "ibamr/INSStaggeredHierarchyIntegrator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
+
+#include "ibtk/CartGridFunction.h"
+#include "ibtk/HierarchyGhostCellInterpolation.h"
+#include "ibtk/ibtk_utilities.h"
+#include "ibtk/muParserCartGridFunction.h"
+#include "ibtk/muParserRobinBcCoefs.h"
+
+#include "Box.h"
+#include "CartesianPatchGeometry.h"
+#include "CellData.h"
+#include "CellIndex.h"
+#include "CellVariable.h"
+#include "HierarchyDataOpsManager.h"
+#include "Index.h"
+#include "IntVector.h"
+#include "Patch.h"
+#include "PatchGeometry.h"
+#include "PatchHierarchy.h"
+#include "PatchLevel.h"
+#include "SideData.h"
+#include "Variable.h"
+#include "VariableContext.h"
+#include "VariableDatabase.h"
+#include "VisItDataWriter.h"
+#include "tbox/Database.h"
+#include "tbox/Pointer.h"
+
+#include <Eigen/Cholesky>
+#include <Eigen/Core>
+#include <Eigen/Eigenvalues>
+#include <unsupported/Eigen/MatrixFunctions>
+
+#include <math.h>
+#include <mpi.h>
+
+#include <string>
+#include <vector>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class ComplexFluidForcing provides an interface for specifying a viscoelastic stress to be added to the
```

Comment:
Would consider `PolymericStressForcing` or `ExtraStressForcing`.

Reply:
Also, for classes that require a particular data centering, we usually put the data centering in the class name. (Not sure if this is the best thing to do or not.)

Reply:
This works for both the cell centered and side centered fluid solver. The only difference is the stencil for the divergence of the conformation tensor.

Reply:
But the conformation tensor is required to be cell centered?

Reply:
Yes, but only because the advection diffusion solver requires it.

I'm not sure how it should be named. All the INS forcing functions are named based on which solver they work with.